pkg/streamingpromql/operators/functions/histograms.go

```diff
@@ -0,0 +1,346 @@
+	nativeHistograms     *[]promql.HPoint // Histograms should only ever exist once per group
+	remainingSeriesCount uint             // The number of series remaining before this group is fully collated.
+}
+
+var bucketGroupPool = zeropool.New(func() *bucketGroup {
+	return &bucketGroup{}
+})
+
+func NewHistogramFunctionOverInstantVector(
+	phArg types.ScalarOperator,
+	inner types.InstantVectorOperator,
+	memoryConsumptionTracker *limiting.MemoryConsumptionTracker,
+	annotations *annotations.Annotations,
+	expressionPosition posrange.PositionRange,
+	timeRange types.QueryTimeRange,
+) *HistogramFunctionOverInstantVector {
+	return &HistogramFunctionOverInstantVector{
+		phArg:                    phArg,
+		inner:                    inner,
+		memoryConsumptionTracker: memoryConsumptionTracker,
+		Annotations:              annotations,
+		innerSeriesMetricNames:   &operators.MetricNames{},
+		expressionPosition:       expressionPosition,
+		timeRange:                timeRange,
+	}
+}
+
+func (h *HistogramFunctionOverInstantVector) ExpressionPosition() posrange.PositionRange {
+	return h.expressionPosition
+}
+
+func (h *HistogramFunctionOverInstantVector) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadata, error) {
+	// Pre-process any Scalar arguments
+	var err error
+	h.phValues, err = h.phArg.GetValues(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	innerSeries, err := h.inner.SeriesMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		types.PutSeriesMetadataSlice(innerSeries)
+	}()
+
+	if len(innerSeries) == 0 {
+		// No input series == no output series.
+		return nil, nil
+	}
+
+	h.innerSeriesMetricNames.CaptureMetricNames(innerSeries)
+	groups := map[string]groupWithLabels{}
+	h.seriesGroups = make([][]*bucketGroup, 0, len(innerSeries))
+	h.bucketValues, err = types.StringSlicePool.Get(len(innerSeries), h.memoryConsumptionTracker)
+	if err != nil {
+		return nil, err
+	}
+	b := make([]byte, 0, 1024)
+
+	for _, series := range innerSeries {
+		// Each series belongs to two groups
+		seriesGroupPair := make([]*bucketGroup, 2)
+
+		// Store the le label. If it doesn't exist, it'll be an empty string
+		le := series.Labels.Get(labels.BucketLabel)
+		h.bucketValues = append(h.bucketValues, le)
+
+		// First get the group with all labels
+		b = series.Labels.Bytes(b)
+		g, groupExists := groups[string(b)]
+		if !groupExists {
+			lb := labels.NewBuilder(series.Labels)
+			g.labels = lb.Labels()
+			g.group = bucketGroupPool.Get()
+			g.group.groupedMetricName = series.Labels.Get(labels.MetricName)
+			g.group.pointBuckets = make([]buckets, h.timeRange.StepCount)
+			groups[string(b)] = g
+		}
+		g.group.remainingSeriesCount++
+		seriesGroupPair[0] = g.group
+
+		// Now, ignore `le` and create a group with the rest of the labels.
+		// We do this even if we don't have an `le` label. This is so we can detect
+		// where native histograms exist alongside classic histograms.
```

Comment:
Is this necessary? For series without an `le` label, we could just have one group, and set the other element of `seriesGroupPair` to `nil`. If we receive a float value for this series, then we can detect that by checking for that element being `nil` and emit the appropriate annotation.

Reply:
Yep, I'm not sure how much it saves on the series iteration, but it does avoid a lookup in the map.